libcxx/include/fstream

```diff
@@ -158,8 +158,8 @@ public:
     basic_fstream();
     explicit basic_fstream(const char* s, ios_base::openmode mode = ios_base::in|ios_base::out);
     explicit basic_fstream(const string& s, ios_base::openmode mode = ios_base::in|ios_base::out);
-    explicit basic_fstream(const filesystem::path& p,
-                           ios_base::openmode mode = ios_base::in|ios_base::out); C++17
+    template<class T>
+    explicit basic_fstream(const T& s, ios_base::openmode mode = ios_base::in | ios_base::out); // Since C++17
```

Comment:
Actually, https://en.cppreference.com/w/cpp/io/basic_fstream/basic_fstream mentions a `const std::filesystem::path::value_type*` overload which kinda happened to work before this change (since it'd construct a `::path` and call that ctor), but this change breaks that. Should we undo this change here until an explicit `const std::filesystem::path::value_type*` ctor exists?

Reply:
That constructor already exists:

https://github.com/llvm/llvm-project/blob/a417b9b759960cc4d7b49e95ddc2ef129b5abe12/libcxx/include/fstream#L1413-L1415

It's definitely expected that you can't open an fstream with a `wstring`, the constructor aking a `filesystem::path` should not be used to convert "anything string-like, with any character type, in any encoding" into a file name with implicit allocations and character conversions.

Reply:
If you want that conversion, you can still do it explicitly:

```c++
std::fstream model_file(filesystem::path(ml_package_dir()
                              .Append(kMlPackageDataDir)
                              .Append(kMlPackageModelFileName)
                              .value()),
                          std::ios::out | std::ios::binary);
```

Reply:
Oh, I suppose I need to call `c_str()` on my wstring to get the wchar_t ctor. Let me try that; sorry for the noise.

(The point stands that this breaks existing code, though.)

Reply:
It's indeed correct this may break existing code. Louis tested this at Apple, based on the [results](
https://github.com/llvm/llvm-project/pull/85079#issuecomment-2032096059) we applied the possibly breaking change. Is the breakage for you worse?

Reply:
It looks like it's just one file so far (*). I got a bit confused because the wchar_t* ctor is missing from the synopsis comment at the top. So all good.

*: But actually rolling this in is blocked on https://github.com/llvm/llvm-project/pull/86843#issuecomment-2051880399 , so not 100% sure yet. But very likely that's all.